Automatic-Module-Name: Calling all Java Library Maintainers

http://branchandbound.net/blog/java/2017/12/automatic-module-name/

- [x] must fix #179 first 
- [x] must be merged: #315
- [x] must upgrade fop #158 
- [x] https://youtrack.jetbrains.com/issue/IDEA-183692
  - next Intellij IDEA version, expected late July 2022 
- [x] ~https://issues.apache.org/jira/browse/XALANJ-2632~ (This should not be needed with [fop bump](https://github.com/premium-minds/billy/pull/354) that [no longer requires xalan](https://issues.apache.org/jira/browse/FOP-3085))
  - https://stackoverflow.com/questions/54682417/unable-to-derive-module-descriptor-provider-class-x-not-in-module
  - https://github.com/apache/xalan-java/pull/1
  - with java.version set to 11 and Automatic-Module-Name defined in each maven module:
```
$ JAVA_HOME=/usr/lib/jvm/java-1.17.0-openjdk-amd64 mvn -q -DskipTests clean install                                                                                                                 130 ↵
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.4.0:jar (attach-javadocs) on project billy-gin: Execution attach-javadocs of goal org.apache.maven.plugins:maven-javadoc-plugin:3.4.0:jar failed: Unable to derive module descriptor for /home/froque/.m2/repository/xalan/xalan/2.7.2/xalan-2.7.2.jar: Provider class org.apache.bsf.BSFManager not in module -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the '-e' switch
[ERROR] Re-run Maven using the '-X' switch to enable verbose output
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn [args] -r
```

--- a/billy-andorra/pom.xml
+++ b/billy-andorra/pom.xml
@@ -185,6 +185,17 @@
 					</dependency>
 				</dependencies>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.andorra</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
 		</plugins>
 	</build>
 

--- a/billy-core-jpa/pom.xml
+++ b/billy-core-jpa/pom.xml
@@ -116,6 +116,17 @@
 					</execution>
 				</executions>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.persistence</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
 		</plugins>
 	</build>
 

--- a/billy-core/pom.xml
+++ b/billy-core/pom.xml
@@ -91,7 +91,13 @@
 			<plugin>
 				<groupId>org.apache.maven.plugins</groupId>
 				<artifactId>maven-jar-plugin</artifactId>
-				<version>3.3.0</version>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.core</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
 				<executions>
 					<execution>
 						<goals>

--- a/billy-france/pom.xml
+++ b/billy-france/pom.xml
@@ -186,6 +186,17 @@
 					</dependency>
 				</dependencies>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.france</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
 		</plugins>
 	</build>
 

--- a/billy-gin/pom.xml
+++ b/billy-gin/pom.xml
@@ -90,4 +90,20 @@
 		</dependency>
 	</dependencies>
 
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.gin</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
+		</plugins>
+	</build>
+
 </project>

--- a/billy-portugal/pom.xml
+++ b/billy-portugal/pom.xml
@@ -272,6 +272,17 @@
 					<excludePackageNames>com.premiumminds.billy.portugal.services.export.saftpt.*</excludePackageNames>
 				</configuration>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.portugal</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
 		</plugins>
 	</build>
 

--- a/billy-spain/pom.xml
+++ b/billy-spain/pom.xml
@@ -186,6 +186,17 @@
 					</dependency>
 				</dependencies>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-jar-plugin</artifactId>
+				<configuration>
+					<archive>
+						<manifestEntries>
+							<Automatic-Module-Name>com.premiumminds.billy.spain</Automatic-Module-Name>
+						</manifestEntries>
+					</archive>
+				</configuration>
+			</plugin>
 		</plugins>
 	</build>
 

--- a/pom.xml
+++ b/pom.xml
@@ -315,6 +315,11 @@
 						<release>${java.version}</release>
 					</configuration>
 				</plugin>
+				<plugin>
+					<groupId>org.apache.maven.plugins</groupId>
+					<artifactId>maven-jar-plugin</artifactId>
+					<version>3.3.0</version>
+				</plugin>
 				<!--This plugin's configuration is used to store Eclipse m2e settings
 					only. It has no influence on the Maven build itself. -->
 				<plugin>

